Convert buttons to CSS variables

mdo opened the pull request:
**Todos:**

- [x] Replace focus styles with rgb color CSS vars (needs #34100 merged first)
- [x] **Replace or update `gradient()` mixins.** _Nothing to figure out, we're good to go!_
- [x] **Figure out `border-radius` property vs `@include border-radius()`.** _Dropped the mixin here and am using a fallback value in the CSS variable given the `valid-radius()` function._
- [x] Figure out RFS sizing in CSS vars (maybe relevant: https://github.com/twbs/rfs/issues/311). _I've removed the `font-size()` mixin from the `button-size()` mixin and default `.btn` value. It doesn't seem to matter comparing this PR to main._
- [x] Replace or update outline variant mixin and styles.
- [x] Fix `.btn-link`
- [x] **Tighten bundlewatch before the final merge**

---

Biggest concern for all these PRs will be compiled file size of our CSS. Looking at the **bootstrap.css diff** though:

> 319 insertions(+), 485 deletions(-)

~~This will change a bit once `:focus` styles are restored.~~

---

**Preview:** <https://deploy-preview-34600--twbs-bootstrap.netlify.app/docs/5.1/components/buttons/>

mdo committed:
Convert .btn and mixins to use CSS variables

mdo committed:
Tighten bundlewatch

mdo commented:
Per #34435, we may want to migrate this back to v4 with a separate PR to fix in Sass.

mdo commented on site/content/docs/5.1/components/buttons.md:
```diff
@@ -227,13 +238,21 @@ buttons.forEach(function (button) {
 
 ### Variables
 
+<small class="d-inline-flex px-2 py-1 font-monospace text-muted border rounded-3">Added in v5.2.0</small>
```
I'll tackle that in a separate effort—will revamp this and update the other instances.

ffoodd commented on scss/_buttons.scss:
```diff
@@ -1,51 +1,97 @@
 .btn {
+  // scss-docs-start btn-css-vars
+  --#{$variable-prefix}btn-padding: #{$btn-padding-y} #{$btn-padding-x};
+  --#{$variable-prefix}btn-font-family: #{$btn-font-family};
```
Seems unused, isn't it?

ffoodd commented on scss/_buttons.scss:
```diff
@@ -1,51 +1,97 @@
 
   &:hover {
-    color: $body-color;
+    color: var(--#{$variable-prefix}btn-hover-color);
     text-decoration: if($link-hover-decoration == underline, none, null);
+    background-color: var(--#{$variable-prefix}btn-hover-bg);
+    border-color: var(--#{$variable-prefix}btn-hover-border-color);
```
Overriding the existing `--#{$variable-prefix}btn-color` as you did in #34622 is much clearer IMHO. 

However it'd require to keep selectors in the variants' mixin, but I think the overall file size would still be lighter. Any thought?

mdo replied:
While that's maybe a bit cleaner, it has the downside of either doubling some selectors or repeating some rulesets. For example, not all styles of `:focus` or `:disabled` should be in the mixin. Do we duplicate those selectors and have some in `_buttons.scss` and the rest in `mixins/_buttons.scss`?

Sir-Genius commented:
Outline button has no disabled border color.

florianlacreuse commented on scss/_buttons.scss:
```diff
@@ -1,51 +1,97 @@
-  line-height: $btn-line-height;
-  color: $body-color;
+  padding: var(--#{$variable-prefix}btn-padding);
+  font-family: var(--#{$variable-prefix}btn-font-weight);
```
It should be `--#{$variable-prefix}btn-font-family`.

mdo committed:
Rewrite custom docs buttons

- Use CSS variables wherever possible
- Rename purple-bright to violet for brevity, add CSS vars versions
- Rename download to accent

ffoodd commented on scss/_buttons.scss:
```diff
@@ -1,51 +1,97 @@
 
 .btn {
+  // scss-docs-start btn-css-vars
+  --#{$variable-prefix}btn-padding: #{$btn-padding-y} #{$btn-padding-x};
```
Couldn't we add separate custom props for vertical and horizontal padding? I guess this request will pop up for each component.

May even drop `--bs-*-padding` since they're only shorthands, don't you think?

ffoodd commented on scss/_buttons.scss:
```diff
@@ -1,51 +1,97 @@
-  border: $btn-border-width solid transparent;
-  @include button-size($btn-padding-y, $btn-padding-x, $btn-font-size, $btn-border-radius);
+  border: var(--#{$variable-prefix}btn-border-width) solid var(--#{$variable-prefix}btn-border-color);
+  border-radius: var(--#{$variable-prefix}btn-border-radius, 0); // stylelint-disable-line property-disallowed-list
```
Is there a reason to not use the mixin here? Custom property wouldn't work, maybe?

mdo replied:
Had avoided it for some reason but might not be necessary. Will restore it.

ffoodd commented on site/content/docs/5.1/components/buttons.md:
````diff
@@ -227,13 +238,42 @@ buttons.forEach(function (button) {
+  <button type="button" class="btn btn-bd-primary">Custom button</button>
+</div>
+
+```scss
+.btn-bd-primary {
+  --bs-btn-font-weight: 600;
+  --bs-btn-color: var(--bs-white);
+  --bs-btn-bg: var(--bd-violet);
+  --bs-btn-border-color: var(--bd-violet);
+  --bs-btn-hover-color: var(--bs-white);
+  --bs-btn-hover-bg: #{shade-color($bd-violet, 20%)};
+  --bs-btn-hover-border-color: #{shade-color($bd-violet, 20%)};
+  --bs-btn-focus-shadow-rgb: var(--bd-violet-rgb);
+}
+```
````
Could the `scss-docs` shortcode be used, here? Didn't try but could simplify further changes.

ffoodd reviewed (approved):
A few questions around the PR— it looks fine already 👌

Very nice work, I really like the naming move to `accent` in the docs ❤️

mdo commented on scss/mixins/_buttons.scss:
```diff
@@ -1,3 +1,5 @@
+// stylelint-disable custom-property-empty-line-before
```
Agreed, not sure where to update though just yet. 🤔 Open to some direction there, but likely to proceed with this :).
